### PR TITLE
Internal: Recreate prop-types [EDS-532]

### PR DESCRIPTION
**Pull request by Nevoss**

### CODE REVIEW

File: modules/atomic-widgets/prop-types/concerns/has-required-validation.php [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+if ( ! defined( 'ABSPATH' ) ) {
+	exit; // Exit if accessed directly.
+}
+
+/**
+ * @mixin Prop_Type
+ * @mixin Has_Settings
+ */
+trait Has_Required_Validation {
```
Comment (Omerisra6) [inline modules/atomic-widgets/prop-types/concerns/has-required-validation.php:15]: why not Has_Required?

File: modules/atomic-widgets/prop-types/base/array-prop-type.php [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+use Elementor\Modules\AtomicWidgets\PropTypes\Concerns;
+use Elementor\Modules\AtomicWidgets\PropTypes\Contracts\Persistable_Prop_Type;
+use Elementor\Modules\AtomicWidgets\PropTypes\Contracts\Prop_Type;
+
+if ( ! defined( 'ABSPATH' ) ) {
+	exit; // Exit if accessed directly.
+}
+
+abstract class Array_Prop_Type implements Persistable_Prop_Type {
+	const TYPE = 'array';
```
Comment (Omerisra6) [inline modules/atomic-widgets/prop-types/base/array-prop-type.php:14]: We should find better name, maybe data type?

File: modules/atomic-widgets/props-resolver/props-resolver.php [hunk 3]
```diff
@@ -97,10 +89,32 @@ private function transform( $value, int $depth = 0 ) {
...
+		if ( $prop_type instanceof Union_Prop_Type ) {
+			$prop_type = $prop_type->get_prop_type( $value['$$type'] );
+
+			if ( ! $prop_type ) {
+				return null;
+			}
+		}
+
+		if ( $prop_type instanceof Object_Prop_Type ) {
+			if ( ! is_array( $value['value'] ) ) {
```
Comment (Omerisra6) [inline modules/atomic-widgets/props-resolver/props-resolver.php:101]: At that point, isn't the value always an array?

File: modules/atomic-widgets/props-resolver/props-resolver.php [hunk 4]
```diff
@@ -97,10 +89,32 @@ private function transform( $value, int $depth = 0 ) {
...
+			$value['value'] = $this->resolve(
+				$prop_type->get_shape(),
+				$value['value']
+			);
+		}
+
+		if ( $prop_type instanceof Array_Prop_Type ) {
+			if ( ! is_array( $value['value'] ) ) {
+				return null;
+			}
```
Comment (Omerisra6) [inline modules/atomic-widgets/props-resolver/props-resolver.php:114]: same as above

File: modules/atomic-widgets/validators/props-validator.php [hunk 5]
```diff
@@ -35,21 +35,24 @@ public static function make( array $schema ): self {
 	public function validate( array $props ): array {
 		$validated = [];
 
-		foreach ( $props as $key => $value ) {
-			$prop_type = $this->schema[ $key ] ?? null;
-
+		foreach ( $this->schema as $key => $prop_type ) {
```
Comment (Omerisra6) [inline modules/atomic-widgets/validators/props-validator.php:38]: maybe change to prop types as you did in the props resolver?

**Review by Omerisra6 — CHANGES_REQUESTED** [review]

File: modules/atomic-widgets/prop-types/base/array-prop-type.php [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+			'value' => $value,
+		];
+
+		return $this;
+	}
+
+	public function validate( $value ): bool {
+		if ( is_null( $value ) ) {
+			return ! $this->is_required();
+		}
```
Comment (Omerisra6) [inline modules/atomic-widgets/prop-types/base/array-prop-type.php:54]: maybe extract to a trait?

File: modules/atomic-widgets/prop-types/concerns/has-meta.php [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+}
+
+/**
+ * @mixin Prop_Type
+ */
+trait Has_Meta {
+	protected array $meta = [];
+
+	public function meta( $key, $value = null ): self {
+		$is_tuple = is_array( $key ) && 2 === count( $key );
```
Comment (Omerisra6) [inline modules/atomic-widgets/prop-types/concerns/has-meta.php:18]: what is the usage for tuple?

File: tests/phpunit/elementor/modules/atomic-widgets/dynamic-tags/test-dynamic-prop-type.php [hunk 8]
```diff
@@ -37,54 +37,76 @@ public function tearDown(): void {
 		Plugin::$instance->dynamic_tags = $this->original_dynamic_tags;
 	}
```
Comment (Omerisra6) [inline tests/phpunit/elementor/modules/atomic-widgets/dynamic-tags/test-dynamic-prop-type.php:39]: Don't forget to add tests as we discussed

**Review by Omerisra6 — CHANGES_REQUESTED** [review]

File: .grunt-config/webpack.packages.js [hunk 9]
```diff
@@ -61,7 +63,7 @@ const common = {
 const devConfig = {
 	...common,
 	mode: 'development',
-	devtool: false, // TODO: Need to check what to do with source maps.
+	devtool: usingLocalRepo ? 'source-map' : false,
```
[the remark this answers is not in the document]
Comment (Nevoss) [reply .grunt-config/webpack.packages.js:66]: No,  will split it to another PR

File: modules/atomic-widgets/dynamic-tags/dynamic-prop-types-mapping.php [hunk 10]
```diff
@@ -20,51 +27,77 @@ public static function make(): self {
...
 			}
 
-			$this->add_to_prop_type( $prop );
+			$result[ $key ] = $this->get_modified_prop_type( $prop_type );
+		}
+
+		return $result;
+	}
+
+	private function get_modified_prop_type( Prop_Type $prop_type ) {
```
[the remark this answers is not in the document]
Comment (Nevoss) [reply modules/atomic-widgets/dynamic-tags/dynamic-prop-types-mapping.php:46]: add docs, not sure how to extract into method

File: modules/atomic-widgets/dynamic-tags/dynamic-prop-types-mapping.php [hunk 11]
```diff
@@ -20,51 +27,77 @@ public static function make(): self {
...
+
+			$categories = array_merge( $categories, $this->get_related_categories( $p ) );
 		}
 
-		return $schema;
+		if ( empty( $categories ) ) {
+			return $prop_type;
+		}
+
+		$union_prop_type = $prop_type instanceof Persistable_Prop_Type ?
```
[the remark this answers is not in the document]
Comment (Nevoss) [reply modules/atomic-widgets/dynamic-tags/dynamic-prop-types-mapping.php]: Yes it must, a union supports only array of transformable prop types, it cannot create a union inside a union.

File: modules/atomic-widgets/prop-types/base/object-prop-type.php [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+		$this->shape = $shape;
+
+		return $this;
+	}
+
+	public function get_shape(): array {
+		return $this->shape;
+	}
+
+	public function get_shape_item( $key ): ?Prop_Type {
```
[the remark this answers is not in the document]
Comment (Nevoss) [reply modules/atomic-widgets/prop-types/base/object-prop-type.php:54]: Do you have a suggestion?

File: modules/atomic-widgets/prop-types/base/object-prop-type.php [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+		return $this->shape[ $key ] ?? null;
+	}
+
+	public function validate( $value ): bool {
+		if ( is_null( $value ) ) {
+			return ! $this->is_required();
+		}
```
[the remark this answers is not in the document]
Comment (Nevoss) [reply modules/atomic-widgets/prop-types/base/object-prop-type.php:78]: yes we do

File: modules/atomic-widgets/prop-types/concerns/has-required-validation.php (hunk 1 above)
Comment (Nevoss) [reply modules/atomic-widgets/prop-types/concerns/has-required-validation.php:15]: Changed to `Has_Required_Setting` is it better? `Has_Required` feels a little bit weird to me

File: modules/atomic-widgets/prop-types/base/array-prop-type.php (hunk 2 above)
Comment (Nevoss) [reply modules/atomic-widgets/prop-types/base/array-prop-type.php:14]: The problem is that `union` and `plain` is not a real data type, maybe `kind`? or `class`? or `group`? @StyleShit

File: modules/atomic-widgets/props-resolver/props-resolver.php (hunk 3 above)
Comment (Nevoss) [reply modules/atomic-widgets/props-resolver/props-resolver.php:97]: This is for an edge case, when the prop type is an array but the value can be any thing that was successfully saved in the DB, really this is just for protection against failing in the Front End

File: modules/atomic-widgets/props-resolver/props-resolver.php (hunk 4 above)
Comment (Nevoss) [reply modules/atomic-widgets/props-resolver/props-resolver.php:110]: Same

File: modules/atomic-widgets/dynamic-tags/dynamic-prop-type.php [hunk 14]
```diff
@@ -2,21 +2,23 @@
 
 namespace Elementor\Modules\AtomicWidgets\DynamicTags;
 
-use Elementor\Modules\AtomicWidgets\PropTypes\Transformable_Prop_Type;
+use Elementor\Modules\AtomicWidgets\PropTypes\Base\Plain_Prop_Type;
```
[the remark this answers is not in the document]
Comment (Nevoss) [reply modules/atomic-widgets/dynamic-tags/dynamic-prop-type.php:5]: There is a different between `primitive` and `plain`

[the remark this answers is not in the document]
Comment (Nevoss) [reply modules/atomic-widgets/dynamic-tags/dynamic-prop-type.php:5]: Not sure which folder?

File: modules/atomic-widgets/prop-types/base/array-prop-type.php (hunk 6 above)
Comment (Nevoss) [reply modules/atomic-widgets/prop-types/base/array-prop-type.php:67]: I did not find any way to do so... I'm not sure how.

File: modules/atomic-widgets/prop-types/concerns/has-meta.php (hunk 7 above)
Comment (Nevoss) [reply modules/atomic-widgets/prop-types/concerns/has-meta.php:24]:
```php
String_Prop_Type::make()->meta( Dyanmic_Prop_Type::ignore() );
```

the `ignore` method returns `['dynamic', false]` this allows more convenient way for widgets builders to disable dynamic on some prop type

File: tests/phpunit/elementor/modules/atomic-widgets/dynamic-tags/test-dynamic-prop-type.php (hunk 8 above)
Comment (Nevoss) [reply tests/phpunit/elementor/modules/atomic-widgets/dynamic-tags/test-dynamic-prop-type.php:39]: Done

File: modules/atomic-widgets/dynamic-tags/dynamic-prop-types-mapping.php (hunk 11 above)
[the remark this answers is not in the document]
Comment (Omerisra6) [reply modules/atomic-widgets/dynamic-tags/dynamic-prop-types-mapping.php]: So why do we need to check it?

File: modules/atomic-widgets/prop-types/base/object-prop-type.php (hunk 12 above)
[the remark this answers is not in the document]
Comment (Omerisra6) [reply modules/atomic-widgets/prop-types/base/object-prop-type.php:54]: Maybe `get_type_by_key` or `get_shape_field`?
I am trying to treat the shape like a type and not an object...

File: modules/atomic-widgets/prop-types/base/array-prop-type.php (hunk 2 above)
Comment (Omerisra6) [reply modules/atomic-widgets/prop-types/base/array-prop-type.php:14]: `kind` sounds better

File: modules/atomic-widgets/dynamic-tags/dynamic-prop-type.php (hunk 14 above)
[the remark this answers is not in the document]
Comment (Omerisra6) [reply modules/atomic-widgets/dynamic-tags/dynamic-prop-type.php:5]: My bad

File: modules/atomic-widgets/prop-types/base/array-prop-type.php (hunk 6 above)
Comment (Omerisra6) [reply modules/atomic-widgets/prop-types/base/array-prop-type.php:53]: maybe:
```
public function validate_required( $value ) {
    if ( is_null( $value ) ) {
	return ! $this->is_required();
    }
    
    return true;
}

File: modules/atomic-widgets/dynamic-tags/dynamic-prop-types-mapping.php (hunk 11 above)
[the remark this answers is not in the document]
Comment (Nevoss) [reply modules/atomic-widgets/dynamic-tags/dynamic-prop-types-mapping.php]: The original prop type can be union, and I am trying to check if the original is already union, if it is I just add the `Dynamic_Prop_Type`, if not I creating a `Union` and then add the dynamic

**Review by Omerisra6 — DISMISSED** [review]

**Review by Omerisra6 — DISMISSED** [review]

**Review by StyleShit — DISMISSED** [review]

**Review by Omerisra6 — APPROVED** [review]

**Pull request merged**